mm2src/trezor/src/transport/udp.rs

```diff
@@ -0,0 +1,161 @@
+
+        let link = UdpLink::open(&dest).await?;
+        if link.ping().await? {
+            devices.push(UdpAvailableDevice {
+                // model: Model::TrezorEmulator,
+                debug,
+                transport: UdpTransport {
+                    protocol: ProtocolV1 { link },
+                },
+            });
+        }
+        Ok(devices)
+    })
+}
+
+/// An actual serial HID USB link to a device over which bytes can be sent.
+struct UdpLink {
+    pub socket: UdpSocket,
+    pub device: (String, String),
+}
+// No need to implement drop as every member is owned
+
+#[async_trait]
+impl Link for UdpLink {
+    async fn write_chunk(&mut self, chunk: Vec<u8>) -> TrezorResult<()> {
+        debug_assert_eq!(CHUNK_SIZE, chunk.len());
+        io::timeout(Duration::from_millis(WRITE_TIMEOUT_MS), async move {
+            self.socket.send(&chunk).await
+        })
+        .await
+        .map_to_mm(|_e| TrezorError::UnderlyingError(String::from("write timeout")))?;
+        Ok(())
+    }
+
+    async fn read_chunk(&mut self, chunk_len: u32) -> TrezorResult<Vec<u8>> {
+        let mut chunk = vec![0; chunk_len as usize];
+        io::timeout(Duration::from_millis(READ_TIMEOUT_MS), async move {
+            let n = self.socket.recv(&mut chunk).await?;
+            if n == chunk_len as usize {
+                Ok(chunk)
+            } else {
+                Err(io::Error::new(io::ErrorKind::Other, "invalid read size"))
+            }
+        })
+        .await
+        .map_to_mm(|_e| TrezorError::UnderlyingError(String::from("read timeout")))
+    }
+}
+
+impl UdpLink {
+    async fn open(path: &str) -> TrezorResult<UdpLink> {
+        let mut parts = path.split(':');
+        let link = Self {
+            socket: UdpSocket::bind(LOCAL_LISTENER).await?,
+            device: (
+                parts.next().expect("Incorrect Path").to_owned(),
+                parts.next().expect("Incorrect Path").to_owned(),
+            ),
+        };
+        link.socket.connect(path).await?;
+        Ok(link)
+    }
+
+    // Ping the port and compare against expected response
+    async fn ping(&self) -> TrezorResult<bool> {
+        let mut resp = [0; CHUNK_SIZE];
+        self.socket.send("PINGPING".as_bytes()).await?;
+        let size = self.socket.recv(&mut resp).await?;
+        Ok(&resp[..size] == "PONGPONG".as_bytes())
+    }
+}
+
+/// An implementation of the Transport interface for UDP devices.
+// #[derive(Debug)]
+pub struct UdpTransport {
+    protocol: ProtocolV1<UdpLink>,
+}
+
+impl UdpTransport {
+    /// Connect to a device over the UDP transport.
+    async fn connect(device: &UdpAvailableDevice) -> TrezorResult<UdpTransport> {
+        let transport = &device.transport;
+        let mut path = String::new();
+        path.push_str(&transport.protocol.link.device.0);
+        path.push(':');
+        path.push_str(&transport.protocol.link.device.1);
```

Comment:
You can use format here also, if you prefer to keep the module as close as the original, there is no need to do this and some of the above suggestions. But making the functions async instead of `block_on` is needed

Reply:
done